src/components/flowchart/flowchart.js

```diff
@@ -451,32 +495,91 @@ export class FlowChart extends Component {
    */
   handleNodeClick = (event, node) => {
     const { type, id } = node;
+    const { onClickToExpandModularPipeline } = this.props;
+
+    if (type === 'modularPipeline') {
+      onClickToExpandModularPipeline(id);
+    } else {
+      this.handleSingleNodeClick(node);
+
+      // the hold shift only happens on clicking a node first
+      if (event.shiftKey) {
+        this.handleShiftClick(node);
+      }
+    }
+
+    event.stopPropagation();
+  };
+
+  handleSingleNodeClick = (node) => {
+    const { id } = node;
     const {
-      onClickToExpandModularPipeline,
       displayMetadataPanel,
       onLoadNodeData,
-      toSelectedNode,
       onToggleNodeClicked,
+      toSelectedNode,
     } = this.props;
 
-    if (type === 'modularPipeline') {
-      onClickToExpandModularPipeline(id);
+    displayMetadataPanel ? onLoadNodeData(id) : onToggleNodeClicked(id);
+    toSelectedNode(node);
+
+    const { from, to, range } = this.state.slicedPipelineState;
+    // if both "from" and "to" are defined
+    // then on a single node click, it should reset the sliced pipeline state
+    if (from !== null && to !== null) {
+      this.updateSlicedPipelineState(null, null, []);
     } else {
-      displayMetadataPanel ? onLoadNodeData(id) : onToggleNodeClicked(id);
-      toSelectedNode(node);
+      // Else, set the first node as the 'from' node based on current state
+      // we need this so that if user hold shift and click on a second node,
+      // the 'from' node is already set
+      this.updateSlicedPipelineState(id, to, range);
     }
-    event.stopPropagation();
+
+    // Reset the slicePipeline on single node click
+    this.props.onSlicePipeline(null, null);
+  };
+
+  handleShiftClick = (node) => {
+    // Close meta data panel
+    this.props.onLoadNodeData(null);
+
+    const fromNodeId = this.state.slicedPipelineState.from || node.id;
+    const toNodeId = node.id;
+
+    const newState = {
+      ...this.state.slicedPipelineState,
+      from: fromNodeId,
+      to: toNodeId,
+    };
+
+    this.setState({ slicedPipelineState: newState });
+
+    this.props.onSlicePipeline(fromNodeId, toNodeId);
+    this.props.onApplySlice(false);
   };
 
   /**
    * Remove a node's focus state and dim linked nodes
    */
-  handleChartClick = () => {
+  handleChartClick = (event) => {
     if (this.props.clickedNode) {
       this.props.onLoadNodeData(null);
       // To reset URL to current active pipeline when click outside of a node on flowchart
       this.props.toSelectedPipeline();
     }
+
+    // Determine if the click event occurred on the slice button.
+    const isSliceButtonClicked =
+      this.slicedPipelineActionRef.current &&
+      this.slicedPipelineActionRef.current.contains(event.target);
```

Comment:
Any specific reason why we are using `ref` to determine click?  how about `onClick`  on `SlicedPipelineActionBar`?

Reply:
hey because the way the flowchart setup, any click outside the element will be considered as "click outside" then it will clear the flowchart. We don't want it to reset everything when clicking on the slicing action (eg: copy command line, click slice button". So this ref is for that reason, hope it makes sense

Reply:
@Huongg 

In `flowchart.js`

```
  onSlicePipelineClick = (event) => {
    console.log('Slice pipeline clicked', event);
  };
  
  ...
  
<div ref={this.slicedPipelineActionRef}>
  <SlicedPipelineActionBar
    chartSize={chartSize}
    onSlicePipelineClick={this.onSlicePipelineClick}
    slicedPipeline={slicedPipelineState.range}
  />
</div>
```

In `sliced-pipeline-action-bar.js`

```
export const SlicedPipelineActionBar = ({ chartSize, slicedPipeline, onSlicePipelineClick }) => {

...

<div className="sliced-pipeline-action-bar--cta">
  <Button onClick={e => onSlicePipelineClick(e)}>Slice</Button>
</div>
```

I did the above changes in you branch and I clicked in `Slice` button and this does not make "click outside" flowchart. Did you tried this? or I am missing something?

Reply:
@Huongg Yes I got it, after removing `ref` it clears the chart. but we can fix this by simply adding `event.stopPropagation();` in `onSlicePipelineClick` in `flowchart.js` to stop click event propagation to parent i our case its flowchart.

```
  onSlicePipelineClick = (event) => {
    event.stopPropagation();
    console.log('Slice pipeline clicked', event);
  }; 
```

Reply:
@Huongg 
In the `SlicedPipelineActionBar` component, if you click on Slice button its already handled and other area like anywhere or on '25 selected', nothing will happen as there is no click event attached, and if you add anything to `SlicedPipelineActionBar` like the run command component or copy button you can alway add `stopPropagation ` to click handler.

The only problem with `ref` approach is its not straight forward  and `onClick` is simple and easy to understand. let me know what do you think.

<img width="259" alt="Screenshot 2024-07-29 at 5 13 41 p m" src="https://github.com/user-attachments/assets/a6d6dcbe-4b36-4d5b-9181-31e073cd2386">

Reply:
hey @jitu5 lets have a chat as maybe it's easier. 
I've attached a screen recording of other area would also need to the `event.stopPropagation` if we don't use the ref approach, especially then `CommandCopier` as there's a lot of things happened there and it doesn't always have the event passing around. 

For me it's easier to handle this in flowchart as it's the parent component, rather than `event.stopPropagation` in other places like ActionBar, CommandCopier, and there might be other places too. 


https://github.com/user-attachments/assets/62a37bf3-8f0d-40c8-8a5e-753cd710fd36

Reply:
I'm not very familiar with the issue but what would be the downside if we use `ref`?

Reply:
@jitu5 and I agree that let's keep the ref for now, and we will revisit it again in the second PR with the run command and everything in the ActionBar